### PR TITLE
add pureconfig module and typesafe configvalue descriptor

### PR DESCRIPTION
**Pull request by yangzai**
Derives `ConfigDescriptor` from Pureconfig `ConfigConvert`.

This requires a `ConfigValue` source descriptor as `ConfigConvert` provides the necessary transformation to and from `ConfigValue`.

Will add more test cases and docs if the general concept is OK.

### CODE REVIEW

File: typesafe/shared/src/main/scala/zio/config/typesafe/TypesafePropertyType.scala
Comment (yangzai) [inline typesafe/shared/src/main/scala/zio/config/typesafe/TypesafePropertyType.scala:13]: I couldn't find an accessible `ConfigValue` parser so this is kind of a hack. First we try parsing the string as a `Config` and if it succeeds we extract the `ConfigObject` within the `Config`. If we fail we prepend a key so it essentially gets wrapped into an object and we try parsing it again. If it succeeds we extract the value with the key.

The reason we don't skip the first parse is that because a hocon object might not be wrapped within parenthesis and prepend a key to it would produce an invalid string for the second parse.

**Comment by afsalthaj** [comment]
I will take a look

**Comment by yangzai** [comment]
@afsalthaj I fixed some CI issues and it needs approval again

**Review by afsalthaj — APPROVED** [review]

**Comment by afsalthaj** [comment]
@yangzai  you might need to wait for this PR to get merged.
https://github.com/zio/zio-config/pull/617

**Comment by afsalthaj** [comment]
While almost all of the pure-config way of derivation is supported in zio-config, I believe this change is still invaluable especially for quick migrations. Once again thanks for the PR, and I will be merging this as soon as the build is ready (I have updated again), and schedule for the next release.

File: typesafe/shared/src/main/scala/zio/config/typesafe/TypesafeConfigDescriptor.scala
Comment (afsalthaj) [inline typesafe/shared/src/main/scala/zio/config/typesafe/TypesafeConfigDescriptor.scala:6]: Hey could we rename this to PureConfigDescriptor ?

File: typesafe/shared/src/main/scala/zio/config/typesafe/TypesafePropertyType.scala
Comment (afsalthaj) [inline typesafe/shared/src/main/scala/zio/config/typesafe/TypesafePropertyType.scala:10]: Could we rename to ConfigValuePropertyType?

**Comment by afsalthaj** [comment]
Sorry, I realised later, that a few renaming is important, just to stop confusing users between what's provided in zio-config-typesafe and this module.

File: typesafe/shared/src/main/scala/zio/config/typesafe/TypesafePropertyType.scala
Comment (yangzai) [reply typesafe/shared/src/main/scala/zio/config/typesafe/TypesafePropertyType.scala:10]: @afsalthaj but the inner object is already `ConfigValuePropertyType`. We could drop the outer `TypesafePropertyType`.

File: typesafe/shared/src/main/scala/zio/config/typesafe/TypesafeConfigDescriptor.scala
Comment (yangzai) [reply typesafe/shared/src/main/scala/zio/config/typesafe/TypesafeConfigDescriptor.scala:6]: @afsalthaj how about we just move the inner `configValueConfigDescriptor` into package and drop `TypesafeConfigDescriptor`?

File: typesafe/shared/src/main/scala/zio/config/typesafe/TypesafePropertyType.scala
Comment (afsalthaj) [reply typesafe/shared/src/main/scala/zio/config/typesafe/TypesafePropertyType.scala:10]: Makes sense.

**Comment by afsalthaj** [comment]
Waiting for this change to finally merge for our next release.

**Review by afsalthaj — APPROVED** [review]

**Pull request merged**
